src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/SecurityServiceDispatcher.cs

```diff
@@ -0,0 +1,593 @@
+
+            //    soapBindingFactory.ProtectionRequirements.IncomingSignatureParts.AddParts(bodyPart, this.IssueAction);
+            //    soapBindingFactory.ProtectionRequirements.IncomingEncryptionParts.AddParts(bodyPart, this.IssueAction);
+            //    soapBindingFactory.ProtectionRequirements.IncomingSignatureParts.AddParts(bodyPart, this.RenewAction);
+            //    soapBindingFactory.ProtectionRequirements.IncomingEncryptionParts.AddParts(bodyPart, this.RenewAction);
+            //}
+
+            //TODO renew part
+            //SupportingTokenParameters renewSupportingTokenParameters = new SupportingTokenParameters();
+            //SecurityContextSecurityTokenParameters sctParameters = new SecurityContextSecurityTokenParameters();
+            //if (bootStrapSecurity.EndpointSupportingTokenParameters.Endorsing.Count > 0)
+            //{
+            //    SecureConversationSecurityTokenParameters scParametersTemp = bootStrapSecurity.EndpointSupportingTokenParameters.Endorsing[0] as SecureConversationSecurityTokenParameters;
+            //    sctParameters.RequireDerivedKeys = scParametersTemp.RequireDerivedKeys;
+            //}
+            //renewSupportingTokenParameters.Endorsing.Add(sctParameters);
+           // bootstrapSecurityProtocolFactory.SecurityBindingElement.OperationSupportingTokenParameters.Add(this.RenewAction.Value, renewSupportingTokenParameters);
+            // bootstrapSecurityProtocolFactory.SecurityTokenManager = new SessionRenewSecurityTokenManager(bootstrapSecurityProtocolFactory.SecurityTokenManager, this.sessionTokenAuthenticator, (SecurityTokenResolver)this.IssuedTokenCache);
+            //SecurityChannelListener<TChannel> securityChannelListener = new SecurityChannelListener<TChannel>(
+            //    this.bootstrapSecurityBindingElement, this.IssuerBindingContext);
+            //securityChannelListener.SecurityProtocolFactory = bootstrapSecurityProtocolFactory;
+            //securityChannelListener.SendUnsecuredFaults = !SecurityUtils.IsCompositeDuplexBinding(context);
+
+            //ChannelBuilder channelBuilder = new ChannelBuilder(context, true);
+            //securityChannelListener.InitializeListener(channelBuilder);
+            //this.shouldMatchRstWithEndpointFilter = SecurityUtils.ShouldMatchRstWithEndpointFilter(this.bootstrapSecurityBindingElement);
+            //return securityChannelListener;
+            bootstrapSecurityProtocolFactory.OpenAsync(TimeSpan.Zero);
+            base.SecurityProtocol = bootstrapSecurityProtocolFactory.CreateSecurityProtocol(
+                    null,
+                    null,null,
+                    true, TimeSpan.Zero);
+
+        }
+
+        public Task OpenAsync(CancellationToken token)
+        {
+            return Task.CompletedTask;
+        }
+
+        internal RequestContext ProcessReceivedRequest(RequestContext requestContext)
+        {
```

Comment:
Thinking this through about the flow of requests when doing Message Security, in WCF it would be a channel which the top level ChannelHandler would call ReceiveRequest (or some variant) on which deferred down the chain to inner channels until a message came out. That message would then bubble up the layers, some work would be done at some layers, some layers would short-circuit the call and actually send the response etc. But in CoreWCF a Channel doesn't have a method that's called on it to get the message, it's pushed up through the IServiceChannelDispatcher.DispatchAsync method. The Channel object has now really become delegated to a holder of state information as well as providing a method to send a response on the transport. So the function of IChannel has been split into two. IServiceChannelDispatcher now provides the functional chaining that IReplyChannel.ReceiveRequest used to do and Channel just holds state and a way to reply. I think we will still sometimes create a new Channel object when we do things like create a non-transport session (NetTcp has a transport session tied to the socket, but WSHttpBinding has a session not tied to a single socket). So with this work, when a session is created when an SCT token is issued, then we will create a Channel object stored in a session cache so when a Channel comes up the line with a request, we will match the Channel and use that for the onward dispatch. But that's to hold state.  
Basically what I'm saying is at this layer we don't need different Channel object's for different channel shapes, we need different IServiceChannelDispatcher implementations. So convert SecurityReplyChannel into a SecurityReplyChannelDispatcher which holds a reference to the incoming transport Channel. Then this work in ProcessReceivedRequest will get called from DispatchAsync.